### ISSUE TITLE
Training YoloV3

### ISSUE THREAD

**Issue by datatemb** [PacktPublishing/Hands-On-Computer-Vision-with-Tensorflow#13]
When training a tf 2.0 YoloV3 model Chapter 5 mentions training  using VGG16 or other pretrained NN as a backbone after the head is deactivated. How do you format the y_true value i.e. (batch_size, grid, grid, anchors, (x1, y1, x2, y2, obj, cls)). Batch size is obvious but I wanted to see an example of how the rest would look and how I should format the data to use it in a model.fit operation, for example. I have annotated the images and have the x and y values for the bounding boxes and obviously the class, but I'm not sure how to supply obj value ( and all the other values when transitioning from the backbone to the YoloV3 head). Is obj (objectness) always going to be 1 for y_true since there is always an object in the bounding box for y_true? Is the class simply going to be the corresponding text label used in the coco.names file or some other format. Is there a way to view a full example of the proposed implemention i.e VGG then YoloV3 for example.

**Comment by EliotAndres** [comment]
Hello, 
Thank you for your question. First of all, `obj` can't always be 1 as otherwise the network would always predict 1.

When generating the data, we start from the labels. For instance, if an image contains 2 objects, there will be 2 labels for this image. However, the Yolo model will predict N = `grid * grid * num_anchors` boxes! So how do we match them?

Well, in practice this is done using the indicator function of the loss. For each label, we assign one box, called "responsible" for this label: 
<img width="684" alt="image" src="https://user-images.githubusercontent.com/9083711/73013629-92e08880-3dcd-11ea-8e69-d39f92af6303.png">

So to answer your question, objectness will be 1 only if the anchor box has the highest IoU with one of the labels.

Regarding the implementation, I highly recommend this implementation: https://github.com/zzh8829/yolov3-tf2 It's very clean and easy to read.

Let me know if you have another question.

**Comment by datatemb** [comment]
Great. Thank you very much !!

So in the yolov3-tf2.py file, lines 33 to 38 define the anchors and anchor masks.

What specifically do those represent? Is each tuple in the array  just a pixel height and width values for each anchor box, normalized to work with a normalized image during processing , i.e. divided by the image size of 416 ? And from my understanding, masks define the order in which to process them, in this case largest to smallest for both yolo and tiny yolo?

Best regards,

Edwin


Sent from Outlook
________________________________

**Comment by EliotAndres** [comment]
Sorry about the late answer. 

Regarding anchor boxes, here's the explanation from the book. If you have specific questions let me know:
<img width="741" alt="image" src="https://user-images.githubusercontent.com/9083711/73645389-8256ca80-4677-11ea-933c-6bd24ec3f689.png">
<img width="759" alt="image" src="https://user-images.githubusercontent.com/9083711/73645404-8a166f00-4677-11ea-9a1b-dc8f5012b13f.png">

Regarding the anchor masks, this is something new in YOLO v3. A good explanation is available here: https://github.com/pjreddie/darknet/issues/558
